homeassistant/components/upnp/device.py

```diff
@@ -96,6 +102,11 @@ def device_type(self) -> str:
         """Get the device type."""
         return self._igd_device.device_type
 
+    @property
+    def unique_id(self) -> str:
+        """Get the unique id."""
+        return f"{self.udn}::{self.device_type}"
```

Comment:
Shouldn't this include ST?

Reply:
This is confusing. The SSDP protocol provides a `USN` (`UDN` and `ST` combined.) E.g., a search-response is something like (converted to JSON, as the CLI for `async_upnp_client` presents it like so, keys prefixes with `_` added by library):
```
{"Cache-Control": "max-age=1900", "Location": "http://192.168.178.1:80/RootDevice.xml", "Server": "UPnP/1.0 UPnP/1.0 UPnP-Device-Host/1.0", "ST": "urn:schemas-upnp-org:device:InternetGatewayDevice:1", "USN": "uuid:upnp-InternetGatewayDevice-1_0-889ffacb9043::urn:schemas-upnp-org:device:InternetGatewayDevice:1", "EXT": "", "_timestamp": "2020-04-27 21:33:01.811349", "_address": "192.168.178.1:1900", "_udn": "uuid:upnp-InternetGatewayDevice-1_0-889ffacb9043", "_source": "search"}
```

The UPnP XML description provides a `UDN` and a `deviceType` (together these form the `USN` from the SSDP protocol), e.g.:
```
<?xml version="1.0" encoding="utf-8"?>
<root xmlns="urn:schemas-upnp-org:device-1-0">
  ...
  <device>
    <deviceType>urn:schemas-upnp-org:device:InternetGatewayDevice:1</deviceType>
    <UDN>uuid:upnp-InternetGatewayDevice-1_0-889ffacb9043</UDN>
    ...
  </device>
</root>
```

Some criticism from my side: The SSDP component in hass returns the XML (in dict form), but entirely discards the search-information. Can be easily worked around, but requires a bit of additional work.

Reply:
Ok.

Feel free to update the SSDP integration 👍 (although we use netdisco for ssdp scan, which has been deprecated…)

Reply:
I couldn't find the required information in the current implementation. It probably requires a change to netdisco.

What do you propose to do with SSDP? Rewrite it to drop netdisco? Should it use its own search implementation or use a library for this? (`async_upnp_client` provides this, which is what the `upnp` component uses.) Also, devices also send out advertisements via SSDP, which home assistant could use to passively listen for, instead of 'actively' searching for devices.

Reply:
Yeah, `async_upnp_client` and passive event listening sound good 👍 `netdisco` has been deprecated and should be dropped.